### ISSUE TITLE
Binding to non-null IEnumerable doesn't work

### ISSUE THREAD

**Issue by brockallen** [dotnet/runtime#36390]
[AB#1277574](https://devdiv.visualstudio.com/10e66e43-9645-4201-b128-0fdc3769cc17/_workitems/edit/1277574)
When using the binding feature in the config system, for IEnumerable collections that have a default, non-null value, the collection is never assigned into the target model. For example:

For this json:
```
{
  "config": {
    "data": [ "a", "b" ]
  }
}
```

and this application code:

```
 public class Config
    {
        IEnumerable<string> _data = new List<string>();
        public IEnumerable<string> Data
        {
            get
            {
                return _data;
            }
            set
            {
                _data = value;
            }
        }
    }

    public class Program
    {
        public static void Main(string[] args)
        {
            var config = new ConfigurationBuilder()
                .SetBasePath(Directory.GetCurrentDirectory())
                .AddJsonFile("test.json")
                .Build();

            var val = config.GetSection("config").Get<Config>();
            Console.WriteLine(val?.Data?.Count());
        }
    }
```

The output is:

```
0
Press any key to continue . . .
```

I'd expect it to print "2".

I'd argue the issue is in `BindInstance` and explicit handling for `IEnumerable` should be made (much like the check for `IsArray` https://github.com/aspnet/Configuration/blob/dev/src/Microsoft.Extensions.Configuration.Binder/ConfigurationBinder.cs#L281).

**Comment by IanKemp** [comment]
For .NET 5.0: https://github.com/dotnet/runtime/blob/master/src/libraries/Microsoft.Extensions.Configuration.Binder/src/ConfigurationBinder.cs#L327
For .NET Core 3.1.8: https://github.com/dotnet/extensions/blob/v3.1.8/src/Configuration/Config.Binder/src/ConfigurationBinder.cs#L328

@lbogdanov basically this is never going to be fixed because the team doesn't care enough to do it. Instead of fixing bugs they would rather just keep piling on new features, because bugs are boring and don't get blog posts written about them.

**Comment by reacheight** [comment]
Basically, why don't we check if the instance runtime type is some `ICollection<T>` instead of checking its declared type in https://github.com/dotnet/runtime/blob/master/src/libraries/Microsoft.Extensions.Configuration.Binder/src/ConfigurationBinder.cs#L334? If it is, then its declared type is either some `ICollection<T>` or `IEnumerable<T>`, and binding should be fine.
```c#
collectionInterface = FindOpenGenericInterface(typeof(ICollection<>), instance.GetType());
if (collectionInterface != null)
{
     BindCollection(instance, collectionInterface, config, options);
}
```
I've checked this on both empty and not empty initialized `IEnumerable<T>`s and it seems to work, all tests are also passed. If that's okay, I'd like to open PR with this change.

**Comment by ericstj** [comment]
It seems reasonable to treat an empty enumerable in the same way as null.  I'm less inclined to examine the runtime type unless we have some precedent for that.  Additionally, it doesn't fix the cases where folks backed the property with Enumerable.Empty or an Array.  cc @eerhardt @maryamariyan

**Comment by ericstj** [comment]
Related: https://github.com/dotnet/runtime/issues/46988 (which could also solve this problem)

**Comment by maryamariyan** [comment]
As part of the closing discussions for issue #46988, it was noted that the current behavior for collections seems unexpected and rather than introducing an opt-in API for it, we could perhaps fix the default behavior. Using this issue to track the fix.

cc: @halter73 @safern

**Comment by SteveDunn** [comment]
I'd like to look at this if it's still 'up-for-grabs'?

**Comment by ericstj** [comment]
I believe it's still up for grabs, and something that we'd like to fix for 6.0.

**Comment by SteveDunn** [comment]
Excellent! I'm very keen to help!

**Comment by safern** [comment]
Thanks, @SteveDunn, I've assigned you the issue. Let me know how I can help.

**Comment by SteveDunn** [comment]
@safern thank you for your offer - I will take you up on your kind offer of help!  Any idea re. this? https://discord.com/channels/732297728826277939/732297728826277942/839396094618238996

This is my first PR on this repo and I'm stumbling around like a newborn Giraffe!

**Comment by safern** [comment]
> This is my first PR on this repo and I'm stumbling around like a newborn Giraffe!

Not a problem. I've pinged you in discord, also feel free to post your questions here if you rather keep them in the open.

**Comment by danmoseley** [comment]
@maryamariyan I'm going through 1st party asks. From what I can see, this is still open, we're still interested in doing this, but it can't fit this release any more? 
If so please change milestone. Or close if appropriate.

**Comment by maryamariyan** [comment]
Will discuss with @halter73 on next steps for this.

**Comment by fakhrulhilal** [comment]
I have the same problem, and use this workaround to avoid analyzer complaint:
```cs
public class Config
{
    public IEnumerable<string> Data { get; set; } = null!;
}
```
But I agree to solve this. No one would understand to make it work without complain from analyzer.

**Comment by SteveDunn** [comment]
> I have the same problem, and use this workaround to avoid analyzer complaint:
> 
> ```cs
> public class Config
> {
>     public IEnumerable<string> Data { get; set; } = null!;
> }
> ```
> 
> But I agree to solve this. No one would understand to make it work without complain from analyzer.

@maryamariyan - do you know if there's been any more thought on this one?

**Comment by maryamariyan** [comment]
Here's a summary of where we're at with this issue:

When we set this issue up for grabs the first time, the desired fix was not clearly defined.

As part of closing the original PR we uncovered that there is a subtle difference between configuration binding and System.Text.Json deserialization. Because with configuration we have to append data from multiple sources but deserialization has a single source of truth and that is why we don't think of configuration binding as a deserializer. (In a way for the configuration case we are mapping what is already been de-serialized to a POCO.)

This comment covers where we left off from the original PR: https://github.com/dotnet/runtime/pull/52514#discussion_r635612319

We decided since then that the proper fix here for this issue would be very similar to `BindArray`: https://github.com/dotnet/runtime/blob/0ee10e98daa8ec07f34a2f9daafee965dd0e80eb/src/libraries/Microsoft.Extensions.Configuration.Binder/src/ConfigurationBinder.cs#L449-L482.

@SteveDunn we could set it up for grabs if you're open to making the above fix. Thanks again for staying engaged on this.

/cc @halter73 @davidfowl
